+bids/layout.m

```diff
@@ -966,243 +257,44 @@ function tolerant_message(tolerant, msg)
-%
 
 function file_list = return_file_list(modality, subject)
-
-  switch modality
-
-    case 'anat'
-      pattern = '_([a-zA-Z0-9]+){1}\\.nii(\\.gz)?';
-
-    case 'func'
-      pattern = '_task-.*_bold\\.nii(\\.gz)?';
-
-    case 'fmap'
-      pattern = '\\.nii(\\.gz)?';
-
-    case 'eeg'
-      pattern = '_task-.*_eeg\\..*[^json]';
-
-    case 'meg'
-      pattern = '_task-.*_meg\\..*[^json]';
-
-    case 'beh'
-      pattern = '_(events\\.tsv|beh\\.json|physio\\.tsv\\.gz|stim\\.tsv\\.gz)';
-
-    case 'dwi'
-      pattern = '_([a-zA-Z0-9]+){1}\\.nii(\\.gz)?';
-
-    case 'pet'
-      pattern = '_task-.*_pet\\.nii(\\.gz)?';
-
-    case 'ieeg'
-      pattern = '_task-.*_ieeg\\..*[^json]';
-
-    case 'perf'
-      pattern = '_(asl|m0scan)\\.nii(\\.gz)?';
-
-  end
-
-  pth = fullfile(subject.path, modality);
-
-  [file_list, d] = bids.internal.file_utils('List', ...
-                                            pth, ...
-                                            sprintf(['^%s.*' pattern '$'], ...
-                                                    subject.name));
-
-  if strcmp(modality, 'meg') && isempty(file_list)
-    file_list = d;
-  end
-
-  file_list = convert_to_cell(file_list);
-
-end
-
-function file_list = return_event_file_list(modality, subject)
-  %
-  % TODO: events file can also be stored at higher levels (inheritance principle)
-  %
-
-  switch modality
-
-    case {'func', 'eeg', 'meg'}
-      pattern = '_task-.*_events\\.tsv';
-
+  if isempty(subject.session)
+    prefix = sprintf('%s_', subject.name);
+  else
+    prefix = sprintf('%s_%s_', subject.name, subject.session);
   end
 
-  pth = fullfile(subject.path, modality);
-
-  [file_list, d] = bids.internal.file_utils('List', ...
-                                            pth, ...
-                                            sprintf(['^%s.*' pattern '$'], ...
-                                                    subject.name));
-
-  file_list = convert_to_cell(file_list);
-
-end
-
-function file_list = return_physio_stim_file_list(modality, subject)
-  %
-  % Physiological and other continuous recordings file
-  %
-  % TODO: stim files can also be stored at higher levels (inheritance principle)
-  %
-
-  switch modality
-
-    case {'func'}
-      pattern = '_task-.*_(physio|stim)\\.tsv\\.gz';
-
-  end
+  % common pattern for all modalities
+  pattern = sprintf(['^%s'... % subject & session
+                     '([a-zA-Z0-9]+-[a-zA-Z0-9]+_)*'... % entities
+                     '([a-zA-Z0-9]+){1}'... % suffix
+                     '\\.(?!json)[a-zA-Z0-9.]+$'... % extention
+                     ], prefix);
 
   pth = fullfile(subject.path, modality);
 
   [file_list, d] = bids.internal.file_utils('List', ...
                                             pth, ...
-                                            sprintf(['^%s.*' pattern '$'], ...
-                                                    subject.name));
-
+                                            pattern);
```

Comment:
You are still going to need to have something like this here because some of the "files" for meg are actually folders.

```
  if strcmp(modality, 'meg') && ~isempty(d)
    for i = 1:size(d, 1)
      file_list{end + 1, 1} = d(i, :);
    end
  end
```